# TensorFlow Assignment: Multilayer Perceptron (MLP) and Convolutional Neural Network (CNN)

**[Duke Community Standard](http://integrity.duke.edu/standard.html): By typing your name below, you are certifying that you have adhered to the Duke Community Standard in completing this assignment.**

Name: [Ziyi Zhang]

Now that you've run through a simple logistic regression model on MNIST, let's see if we can do better (Hint: we can). For this assignment, you'll build a multilayer perceptron (MLP) and a convolutional neural network (CNN), two popular types of neural networks, and compare their performance. Some potentially useful code:

In [1]:
import tensorflow as tf
from tensorflow.examples.tutorials.mnist import input_data

# Import data
mnist = input_data.read_data_sets("MNIST_data/", one_hot=True)

Successfully downloaded train-images-idx3-ubyte.gz 9912422 bytes.
Extracting MNIST_data/train-images-idx3-ubyte.gz
Successfully downloaded train-labels-idx1-ubyte.gz 28881 bytes.
Extracting MNIST_data/train-labels-idx1-ubyte.gz
Successfully downloaded t10k-images-idx3-ubyte.gz 1648877 bytes.
Extracting MNIST_data/t10k-images-idx3-ubyte.gz
Successfully downloaded t10k-labels-idx1-ubyte.gz 4542 bytes.
Extracting MNIST_data/t10k-labels-idx1-ubyte.gz


In [2]:
# Helper functions for creating weight variables
def weight_variable(shape):
    """weight_variable generates a weight variable of a given shape."""
    initial = tf.truncated_normal(shape, stddev=0.1)
    return tf.Variable(initial)

def bias_variable(shape):
    """bias_variable generates a bias variable of a given shape."""
    initial = tf.constant(0.1, shape=shape)
    return tf.Variable(initial)

# Tensorflow Function that might also be of interest:
# tf.nn.sigmoid()
# tf.nn.relu()

### Multilayer Perceptron

Build a multilayer perceptron for MNIST digit classfication. Feel free to play around with the model architecture and see how the training time/performance changes, but to begin, try the following:

Image -> fully connected (500 hidden units) -> nonlinearity (Sigmoid/ReLU) -> fully connected (10 hidden units) -> softmax

Skeleton framework for you to fill in (Code you need to provide is marked by `###`):

In [4]:
# Model Inputs
x = tf.placeholder(tf.float32, [None, 784])
y_ = tf.placeholder(tf.float32, [None, 10])

# Define the graph

# Fully connected 500 layer
W = weight_variable([784, 500])
b = bias_variable([500])
in_image = tf.matmul(x, W) + b
# in_image1 = tf.layers.dense(x,500)

# Nonlinearity layer
# u_1 = tf.nn.relu(in_image)
u_1 = tf.nn.sigmoid(in_image)

# Fully connected 10 layer
W_1 = weight_variable([500,10])
b_1 = bias_variable([10])
hidden_2 = tf.matmul(u_1, W_1) + b_1
    
# Softmax layer
y_mlp = tf.nn.softmax(hidden_2)

# Loss 
cross_entropy = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits(labels=y_, logits=y_mlp))

# Optimizer
train_step = tf.train.GradientDescentOptimizer(0.5).minimize(cross_entropy)

# Evaluation
correct_prediction = tf.equal(tf.argmax(y_mlp, 1), tf.argmax(y_, 1))
accuracy = tf.reduce_mean(tf.cast(correct_prediction, tf.float32))

with tf.Session() as sess:
    # Initialize all variables
    sess.run(tf.global_variables_initializer())
    
    # Training regimen
    for i in range(4000):
        # Validate every 250th batch
        if i % 250 == 0:
            validation_accuracy = 0
            for v in range(10):
                batch = mnist.validation.next_batch(100)
                validation_accuracy += (1/10) * accuracy.eval(feed_dict={x: batch[0], y_: batch[1]})
            print('step %d, validation accuracy %g' % (i, validation_accuracy))
        
        # Train    
        batch = mnist.train.next_batch(50)
        train_step.run(feed_dict={x: batch[0], y_: batch[1]})

    print('test accuracy %g' % accuracy.eval(feed_dict={x: mnist.test.images, y_: mnist.test.labels}))


step 0, validation accuracy 0.068
step 250, validation accuracy 0.693
step 500, validation accuracy 0.724
step 750, validation accuracy 0.748
step 1000, validation accuracy 0.79
step 1250, validation accuracy 0.83
step 1500, validation accuracy 0.792
step 1750, validation accuracy 0.823
step 2000, validation accuracy 0.824
step 2250, validation accuracy 0.817
step 2500, validation accuracy 0.822
step 2750, validation accuracy 0.833
step 3000, validation accuracy 0.822
step 3250, validation accuracy 0.822
step 3500, validation accuracy 0.825
step 3750, validation accuracy 0.837
test accuracy 0.8372


#### Comparison

How do the sigmoid and rectified linear unit (ReLU) compare?

***

for relu, test accuracy is 0.9665
for sigmoid, test accuracy is 0.8372

relu performs better than sigmoid

***

### Convolutional Neural Network

Build a simple 2-layer CNN for MNIST digit classfication. Feel free to play around with the model architecture and see how the training time/performance changes, but to begin, try the following:

Image -> CNN (32 5x5 filters) -> nonlinearity (ReLU) ->  (2x2 max pool) -> CNN (64 5x5 filters) -> nonlinearity (ReLU) -> (2x2 max pool) -> fully connected (1024 hidden units) -> nonlinearity (ReLU) -> fully connected (10 hidden units) -> softmax

Some additional functions that you might find helpful:

In [5]:
# Convolutional neural network functions
def conv2d(x, W):
    """conv2d returns a 2d convolution layer with full stride."""
    return tf.nn.conv2d(x, W, strides=[1, 1, 1, 1], padding='SAME')

def max_pool_2x2(x):
    """max_pool_2x2 downsamples a feature map by 2X."""
    return tf.nn.max_pool(x, ksize=[1, 2, 2, 1], strides=[1, 2, 2, 1], padding='SAME')

# Tensorflow Function that might also be of interest:
# tf.reshape()

Skeleton framework for you to fill in (Code you need to provide is marked by `###`):

*Hint: Convolutional Neural Networks are spatial models. You'll want to transform the flattened MNIST vectors into images for the CNN. Similarly, you might want to flatten it again sometime before you do a softmax. You also might want to look into the  [conv2d() documentation](https://www.tensorflow.org/api_docs/python/tf/nn/conv2d) to see what shape kernel/filter it's expecting.*

In [ ]:
# Model Inputs
x = tf.placeholder(tf.float32, [None, 784])
y_ = tf.placeholder(tf.float32, [None, 10])

# Define the graph

# Reshape image
x_reshaped = tf.reshape(x, shape=[-1, 28, 28, 1])

# CNN 32 & nonlinearity & max pool
Weight_1 = weight_variable([5, 5, 1, 32])
x_1 = max_pool_2x2(tf.nn.relu(conv2d(x_reshaped, Weight_1)))

# CNN 64 & nonlinearity & max pool
Weight_2 = weight_variable([5, 5, 32, 64])
x_2 = max_pool_2x2(tf.nn.relu(conv2d(x_1, Weight_2)))

# Reshape image
x_2 = tf.reshape(x_2, [-1, 7 * 7 * 64])

# fully connected 1024 & nonlinearity
W_1 = weight_variable([7 * 7 * 64,1024])
b_1 = bias_variable([1024])
x_3 = tf.nn.relu(tf.matmul(x_2, W_1) + b_1)

# fully connected 10
W_2 = weight_variable([1024,10])
b_2 = bias_variable([10])
hidden_2 = tf.matmul(x_3, W_2) + b_2

# Loss 
cross_entropy = tf.reduce_mean(tf.nn.softmax_cross_entropy_with_logits(labels=y_, logits=hidden_2))

# Optimizer
train_step = tf.train.AdamOptimizer(1e-4).minimize(cross_entropy)

# Evaluation
correct_prediction = tf.equal(tf.argmax(hidden_2, 1), tf.argmax(y_, 1))
accuracy = tf.reduce_mean(tf.cast(correct_prediction, tf.float32))

with tf.Session() as sess:
    # Initialize all variables
    sess.run(tf.global_variables_initializer())
    
    # Training regimen
    for i in range(10000):
        # Validate every 250th batch
        if i % 250 == 0:
            validation_accuracy = 0
            for v in range(10):
                batch = mnist.validation.next_batch(50)
                validation_accuracy += (1/10) * accuracy.eval(feed_dict={x: batch[0], y_: batch[1]})
            print('step %d, validation accuracy %g' % (i, validation_accuracy))
        
        # Train    
        batch = mnist.train.next_batch(50)
        train_step.run(feed_dict={x: batch[0], y_: batch[1]})

    print('test accuracy %g' % accuracy.eval(feed_dict={x: mnist.test.images, y_: mnist.test.labels}))

step 0, validation accuracy 0.104
step 250, validation accuracy 0.926
step 500, validation accuracy 0.962
step 750, validation accuracy 0.956
step 1000, validation accuracy 0.96
step 1250, validation accuracy 0.978
step 1500, validation accuracy 0.99
step 1750, validation accuracy 0.978
step 2000, validation accuracy 0.986
step 2250, validation accuracy 0.98
step 2500, validation accuracy 0.982
step 2750, validation accuracy 0.98
step 3000, validation accuracy 0.984
step 3250, validation accuracy 0.99
step 3500, validation accuracy 0.986
step 3750, validation accuracy 0.986
step 4000, validation accuracy 0.986
step 4250, validation accuracy 0.992
step 4500, validation accuracy 0.992
step 4750, validation accuracy 0.992
step 5000, validation accuracy 0.988
step 5250, validation accuracy 0.996
step 5500, validation accuracy 0.982
step 5750, validation accuracy 0.98
step 6000, validation accuracy 0.988
step 6250, validation accuracy 0.99
step 6500, validation accuracy 0.984
step 6750, val

Some differences from the logistic regression model to note:

- The CNN model might take a while to train. Depending on your machine, you might expect this to take up to half an hour. If you see your validation performance start to plateau, you can kill the training.

- The logistic regression model we used previously was pretty basic, and as such, we were able to get away with using the GradientDescentOptimizer, which performs implements the gradient descent algorithm. For more difficult optimization spaces (such as the ones deep networks pose), we might want to use more sophisticated algorithms. Prof David Carlson has a lecture on this later.
    
- Because of the larger size of our network, notice that our minibatch size has shrunk.
    
- We've added a validation step every 250 minibatches. This let's us see how our model is doing during the training process, rather than sit around twiddling our thumbs and hoping for the best when training finishes. This becomes especially significant as training regimens start approaching days and weeks in length. Normally, we validate on the entire validation set, but for the sake of time we'll just stick to 10 validation minibatches (500 images) for this homework assignment.

#### Comparison

How do the MLP and CNN compare in accuracy? Training time? Why would you use one vs the other? Is there a problem you see with MLPs when applied to other image datasets?

***

CNN takes longer to train than MLP. CNN has higher accuracy than MLP. For shorter running time, I would use MLP. For higher accuracy, I will use CNN. 

However, when the image is of higher resolution, MLP models will have to flatten the data, which creates an extremely large data. The number of resulting hidden variable will need to be big as well, meaning that we will have to learn many parameters. We cannot ensure the accuracy, nor do we have enough memory to store the parameters.

***